ESQL: Make blocks ref counted

alex-spies opened the pull request:
Make `Block`s reference counted. This allows to replace deep copying of blocks by simply calling `Block::incRef` - the block then has to be closed (or `decRef`ed) one additional time for each call to `incRef` (and `tryIncRef`, if successfull).

alex-spies committed:
Implement references counted block refs

alex-spies commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockRef.java:
```diff
@@ -0,0 +1,78 @@
+
+public class BlockRef implements Releasable {
+
+    private static class RefCount {
```
I'll replace this with a `RefCounted`.

ChrisHegarty replied:
++

alex-spies commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/Block.java:
```diff
@@ -227,7 +227,10 @@ interface Builder extends Releasable {
+     * @deprecated use {@link BlockRef} instead
      */
+    // TODO: Remove this once we use BlockRef
+    @Deprecated
```
This should probably be merged with `BlockRef` instead of deprecating it.

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockRef.java:
```diff
@@ -0,0 +1,78 @@
+
+import org.elasticsearch.core.Releasable;
+
+public class BlockRef implements Releasable {
```
I was thinking you'd stick this in `Block.Ref` actually. It'll change fewer files at least. Not sure that matters though.

nik9000 reviewed (commented):
Right! That's kind of what I was expecting. I suppose the real trick is "what's the least invasive way to get from here to there?"

alex-spies committed:
Use RefCounted

alex-spies committed:
Move BlockRef into Block.Ref

alex-spies committed:
Update tests

alex-spies committed:
Revert all changes

alex-spies committed:
Make Block extend RefCounted

alex-spies committed:
Avoid deep copies in EvalOperator

alex-spies commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/EvalOperator.java:
```diff
@@ -45,7 +44,11 @@ public EvalOperator(BlockFactory blockFactory, ExpressionEvaluator evaluator) {
     protected Page process(Page page) {
         Block.Ref ref = evaluator.eval(page);
-        Block block = ref.floating() ? ref.block() : BlockUtils.deepCopyOf(ref.block(), blockFactory);
+        Block block = ref.block();
+        if (ref.floating() == false) {
+            // We take ownership of this block, so we need to shallow copy (incRef) to avoid double releases.
+            block.incRef();
+        }
```
We can already avoid a deep copy here! :)

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-block-ref-counting

alex-spies committed:
Fix RefCounted javadoc

alex-spies committed:
Improve Block::RefCounted implementation

alex-spies committed:
Add ref counting tests for vector and array blocks

alex-spies committed:
Test remaining block types

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

elasticsearchmachine commented:
Hi @alex-spies, I've created a changelog YAML for you.

alex-spies committed:
Update docs/changelog/100408.yaml

elasticsearchmachine commented:
Pinging @elastic/elasticsearch-esql (:Query Languages/ES|QL)

alex-spies committed:
Fix projection bug, simplify projections

alex-spies commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/Page.java:
```diff
@@ -235,12 +235,8 @@ public void releaseBlocks() {
-            if (map.putIfAbsent(b, Boolean.TRUE) == null) {
-                Releasables.closeExpectNoException(b);
-            }
+            Releasables.closeExpectNoException(b);
```
This becomes simpler to reason about and doesn't need a hashmap on every release.

alex-spies commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/ProjectOperator.java:
```diff
@@ -67,8 +67,10 @@ protected Page process(Page page) {
             var block = page.getBlock(source);
             blocks[b++] = block;
+            block.incRef();
         }
-        return page.newPageAndRelease(blocks);
+        page.releaseBlocks();
+        return new Page(blocks);
```
Another simplification: just close all the blocks and all is good.

alex-spies replied:
Maybe this can be generalized by putting this into the driver, rather than having the operator care about releasing blocks.

alex-spies committed:
Fix projection for empty page

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-block-ref-counting

alex-spies committed:
Remove obsolete test

Filter blocks were removed, the test does not make sense anymore.

alex-spies commented:
@elasticsearchmachine run elasticsearch-ci/part-2

alex-spies commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java:
```diff
@@ -99,6 +97,47 @@ public BlockFactory blockFactory() {
+
+    @Override
+    public final boolean decRef() {
+        close();
```
Feedback from Nhat: make decref do the decrease, then close if count is zero.

alex-spies replied:
Use AbstractRefCounted as template.

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-block-ref-counting

alex-spies committed:
Fix multiply owned BooleanBlock in exchange sink

alex-spies committed:
Improve javadoc and illegal state messages

alex-spies committed:
Remove newPageAndRelease

Can be done easier now by just increasing ref counts of blocks and
releasing the page.

alex-spies committed:
Spotless

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-block-ref-counting

alex-spies committed:
Revert error message

Avoids having to update tests.

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-block-ref-counting

alex-spies committed:
Align logic with AbstractRefCounted

Make decRef() call close once reference count hits 0 instead of having
close() just call decRef().

alex-spies committed:
Cleanup

alex-spies committed:
Address potential partial releases

alex-spies commented:
Hey @dnhatn ! I addressed your comment on aligning `Block` with `AbstractRefCounted`. Let me know what you think!

The contract that I propose now is:
- `decRef` decreases the ref count. If the ref count hits 0, we call `close`. The call to `close` is what actually releases resources, not `decRef`.
- Generally, `decRef` should be used in favor of `close`, but it is still allowed to call `close` iff there is exactly 1 reference (enforced).

I thought about enforcing usage of `decRef` everywhere, but that would kill some nice patterns, esp. try-with-resources in cases when we _know_ that the block was just created and will immediately be discarded again.

Instead, I hunted down the places where `close` is being used and replaced this by `decRef`'ing instead, where applicable.

alex-spies commented:
Also, will update the javadoc in `Block.java` before merging, if we're happy with the updated contract.

alex-spies committed:
Fix test

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java:
```diff
@@ -101,4 +100,54 @@ public BlockFactory blockFactory() {
+        references--;
+
+        if (references <= 0) {
+            close();
```
The consistent pattern is that the `close` method calls the `decRef` method. In turn, the `decRef` method invokes either `closeInternal` or `doClose` when the references reach zero, releasing the memory. This ensures that subclasses can only override `closeInternal` or `doClose`, not the `close` method.

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java:
```diff
@@ -101,4 +100,54 @@ public BlockFactory blockFactory() {
+
+    @Override
+    public void close() {
+        if (released) {
```
While we could enforce a close-once behavior, I prefer allowing multiple-closes to leverage the try-resource.

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/Block.java:
```diff
@@ -281,4 +284,20 @@ static List<NamedWriteableRegistry.Entry> getNamedWriteables() {
+        };
+    }
+
+    static Releasable[] releaseByDecRef(Block... blocks) {
```
I am not sure if we need this. I think I prefer Blocks can be closed multiple times (see the above comment).

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java:
```diff
@@ -101,4 +100,54 @@ public BlockFactory blockFactory() {
+    }
+
+    @Override
+    public void close() {
```
Let's make the close method final.

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java:
```diff
@@ -101,4 +100,54 @@ public BlockFactory blockFactory() {
+    @Override
+    public final void incRef() {
+        if (references <= 0) {
+            throw new IllegalStateException("can't increase refCount on already released block [" + this + "]");
```
Can we add an assertion `assert false : ...` to make tests fail hard?

dnhatn commented on libs/core/src/main/java/org/elasticsearch/core/RefCounted.java:
```diff
@@ -38,7 +38,7 @@ public interface RefCounted {
 
     /**
-     * Tries to increment the refCount of this instance. This method will return {@code true} iff the refCount was
+     * Tries to increment the refCount of this instance. This method will return {@code true} iff the refCount was successfully incremented.
```
Yes, that's correct.

dnhatn reviewed (commented):
I've left some comments, but the PR looks good. I will do another round once you've addressed these comments. Thanks, Alex!

alex-spies committed:
Make decRef and close equivalent again

alex-spies commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java:
```diff
@@ -101,4 +100,54 @@ public BlockFactory blockFactory() {
+    @Override
+    public final boolean decRef() {
+        if (references <= 0) {
+            throw new IllegalStateException("can't release already released block [" + this + "]");
```
Hm, I tried that but it makes many tests awkward as we assert in many tests that an `IllegalStateException` is indeed being thrown.

I would leave out the `assert false ..`; when the ISE is thrown in tests, IME it is not hard to find where it was thrown , and the often used utility `Releasables.expectNoException` internally does `assert false` as well.

alex-spies committed:
Remove releaseByDecRef

alex-spies committed:
Update javadoc for Block

alex-spies commented:
@dnhatn thanks for your review! I implemented your feedback - looking forward to your second round :)

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/X-VectorBlock.java.st:
```diff
@@ -80,15 +83,14 @@ $endif$
     @Override
     public boolean isReleased() {
-        return released || vector.isReleased();
+        return super.isReleased() || vector.isReleased();
```
I think the ownership here is unclear, but it's perfectly fine. Let's leave it as is.

alex-spies replied:
Yeah, I don't like this either.

Currently, the tests do not pass without this, complaining about unclosed blocks. We have a couple of spots where we close vectors directly, but vectors have enclosing blocks (created in the vectors' constructors).

So, IMO; ideally we should refactor at some point so that either
- we always close blocks, not vectors, or
- we stop creating blocks in vectors' constructors - instead, vectors can be completely oblivious to blocks, just being owned by them (my preference).

dnhatn replied:
Thanks, Alex! Yes, I think we can also make Vector ref counted and create a new Block in `asBlock` instead of the constructor. But let's do it later.

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/HashAggregationOperator.java:
```diff
@@ -161,7 +161,7 @@ public void finish() {
             // selected should always be closed
             if (selected != null) {
-                Releasables.closeExpectNoException(selected.asBlock()); // we always close blocks, not vectors
+                selected.asBlock().decRef(); // we always close blocks, not vectors
```
nit: I think we now can close the vector directly here.

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/MvExpandOperator.java:
```diff
@@ -161,7 +161,7 @@ private Page sliceExpandedIntoPages() {
                     prev = null;
                 }
-            }, expandedBlock);
+            }, expandedBlock::decRef);
```
I prefer using `close` here (i.e., just using `expandedBlock`) because we don't explicitly use `incrRef`.

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/MvExpandOperator.java:
```diff
@@ -240,7 +240,11 @@ public void close() {
                 prev.releaseBlocks();
             }
-        }, expandedBlock);
+        }, () -> {
```
Same here, just use `expandedBlock` because we don't explicitly use `incrRef`.

dnhatn commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/LocalExecutionPlanner.java:
```diff
@@ -364,9 +366,11 @@ private PhysicalOperation planExchangeSink(ExchangeSinkExec exchangeSink, LocalE
                     for (int i = 0, s = output.size(); i < s; i++) {
                         var out = output.get(i);
                         if (out.dataType() == DataTypes.BOOLEAN) {
+                            boolBlock.incRef();
```
Should we create a new Block every time here? It's simpler and the cost is trivial (the number of blocks and the cost of ConstantBooleanVector should be small).

alex-spies replied:
Yep, that's even simpler.

dnhatn commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/LocalExecutionPlanner.java:
```diff
@@ -335,8 +335,10 @@ private static Function<Page, Page> alignPageToAttributes(List<Attribute> attrs,
             var blocks = new Block[mappedPosition.length];
             for (int i = 0; i < blocks.length; i++) {
                 blocks[i] = p.getBlock(mappedPosition[i]);
+                blocks[i].incRef();
```
++. This is more correct than the previous version if the same block is mapped to multiple positions.

dnhatn reviewed (approved):
I left some comments, but this looks great. Thanks @alex-spies for the extra iteration.

alex-spies committed:
Address Nhat's feedback

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-block-ref-counting

alex-spies committed:
Align DocBlock and DocVector with others

alex-spies committed:
Add test for closing block's underlying vector

alex-spies committed:
Improve tests

alex-spies committed:
Throw IllegalStateException more consistently

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-block-ref-counting

alex-spies committed:
Check for already released vector only once

alex-spies commented:
@elasticsearchmachine run elasticsearch-ci/part-3

alex-spies committed:
spotless

alex-spies commented:
@elasticsearchmachine run elasticsearch-ci/part-1

alex-spies commented:
@elasticsearchmachine test this please

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-block-ref-counting

alex-spies commented:
@dnhatn , while giving this the final touch I realized that test coverage could have been better.

Now, this is testing all vector/array types in blocks, and we test correct behavior when a vector is closed but the surrounding block is not.

This uncovered small inconsistencies, which I fixed. (`DocBlock` was doing things differently for now good reason; also, ISE was not thrown when closing a block with an already closed vector if the reference count was `>1`.)
